File: shared/yield_curve.py
import os
import csv

import pandas as pd
import polars as pl

from datetime import datetime, date, timedelta
from functools import lru_cache
from typing import Dict, List, Tuple
from dataclasses import dataclass

from options.types.equity import Equity
from shared.constants import dt_fmt_ymd
from shared.paths import Paths

TENOR = 'Tenor'
RATE = 'Rate'


@dataclass
class ZeroCurveData:
    times: List[float]
    rates: List[float]

    def __copy__(self):
        return ZeroCurveData(self.times, self.rates)


class YieldCurve:
    _instance = None
    _initialized = set()

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(YieldCurve, cls).__new__(cls)
            # Maps date (datetime.date) to a dict of {tenor_in_years: zero_rate}
            cls._instance._bill_rates = {}
            cls._instance._par_yields = {}
            cls._instance._zero_curves = {}
        return cls._instance

    def get_file_path(self, market: str, curve_type: str, year: int) -> str:
        return os.path.join(Paths.path_data_interest_rate, market, curve_type, f"{year}.csv")

    def load_bill_rates(self, market: str, dt: date):
        path = self.get_file_path(market, "daily-treasury-bill-rates", dt.year)
        if not os.path.exists(path):
            return

        # Indices for Coupon Equivalent (CE) columns: 2, 4, 6, 8, 10, 12, 14
        ce_indices = [2, 4, 6, 8, 10, 12, 14]
        tenors = [4.0 / 52, 6.0 / 52, 8.0 / 52, 13.0 / 52, 17.0 / 52, 26.0 / 52, 52.0 / 52]

        with open(path, mode='r', encoding='utf-8') as f:
            reader = csv.reader(f)
            next(reader)  # Skip header
            for row in reader:
                try:
                    row_date = datetime.strptime(row[0], "%m/%d/%Y").date()

                    if row_date not in self._bill_rates:
                        self._bill_rates[row_date] = {}

                    for i, idx in enumerate(ce_indices):
                        if idx < len(row) and row[idx]:
                            rate = float(row[idx]) / 100.0
                            self._bill_rates[row_date][tenors[i]] = rate
                except (ValueError, IndexError):
                    continue

    def load_par_yield_curve(self, market: str, dt: date):
        path = self.get_file_path(market, "daily-treasury-par-yield-curve-rates", dt.year)
        if not os.path.exists(path):
            return

        tenors = [1.0 / 12, 1.5 / 12, 2.0 / 12, 3.0 / 12, 4.0 / 12, 6.0 / 12, 1.0, 2.0, 3.0, 5.0, 7.0, 10.0, 20.0, 30.0]

        with open(path, mode='r', encoding='utf-8') as f:
            reader = csv.reader(f)
            next(reader)  # Skip header
            for row in reader:
                try:
                    row_date = datetime.strptime(row[0], "%m/%d/%Y").date()
                    par_yields = []

                    for i, tenor in enumerate(tenors):
                        if (i + 1) < len(row) and row[i + 1]:
                            rate = float(row[i + 1]) / 100.0
                            par_yields.append((tenor, rate))

                    if par_yields:
                        bootstrapped = self.bootstrap_to_zero_rates(par_yields)
                        if row_date not in self._par_yields:
                            self._par_yields[row_date] = {}

                        self._par_yields[row_date].update(bootstrapped)
                except (ValueError, IndexError):
                    continue

    def initialize(self, dt: date, market: str):
        if dt not in self._initialized:
            self._initialized.add(dt)
            self.load_bill_rates(market, dt)
            self.load_par_yield_curve(market, dt)

    @staticmethod
    def has_calibrated_curve(calculation_date: date, equity: Equity = None, market="usa") -> bool:
        if equity is None:
            return False
        root = Paths.path_data_interest_rate / market / 'daily'
        root.mkdir(parents=True, exist_ok=True)
        path = root / f'{equity.symbol.upper()}-{calculation_date.strftime(dt_fmt_ymd)}.csv'
        return path.exists()

    @staticmethod
    def store_calibrated_rates(equity: Equity, calculation_date: date, curve: ZeroCurveData, market="usa"):
        root = Paths.path_data_interest_rate / market / 'daily'
        root.mkdir(parents=True, exist_ok=True)
        path = root / f'{equity.symbol.upper()}-{calculation_date.strftime(dt_fmt_ymd)}.csv'
        df = pl.DataFrame(zip(curve.times, curve.rates), schema=[TENOR, RATE]).cast(
            {TENOR: pl.Float64, RATE: pl.Float64}
        )
        df.write_csv(path)

    @staticmethod
    def load_calibrated_rates(equity: Equity, calculation_date: date, market="usa") -> ZeroCurveData:
        root = Paths.path_data_interest_rate / market / 'daily'
        root.mkdir(parents=True, exist_ok=True)
        path = root / f'{equity.symbol.upper()}-{calculation_date.strftime(dt_fmt_ymd)}.csv'
        if not path.exists():
            raise ValueError(f'load_calibrated_rates(): {path} does not exist.')
        df = pl.read_csv(path)
        return ZeroCurveData(times=list(df.get_column(TENOR)), rates=list(df.get_column(RATE)))

    def get_last_zero_curve(self, calculation_date: date, equity: Equity = None, delta_days=-14, market="usa") -> ZeroCurveData:
        for pd_dt in pd.date_range(calculation_date + timedelta(days=delta_days), calculation_date)[::-1]:
            if self.has_calibrated_curve(pd_dt, equity, market):
                return self.get_zero_curve(pd_dt, equity, market=market)
        return self.get_zero_curve(calculation_date, market=market)

    @lru_cache()
    def get_zero_curve(self, calculation_date: date, equity: Equity = None, market="usa") -> ZeroCurveData:
        if self.has_calibrated_curve(calculation_date, equity, market):
            return self.load_calibrated_rates(equity, calculation_date, market)

        if calculation_date not in self._initialized:
            self.initialize(calculation_date, market)

        if calculation_date not in self._zero_curves:
            bills = self._bill_rates.get(calculation_date, {})
            pars = self._par_yields.get(calculation_date, {})

            if not bills and not pars:
                return ZeroCurveData(times=[], rates=[])

            # Start with bills as zero rates (usually <= 1 year)
            combined_zero_rates = {t: r for t, r in bills.items()}

            combined_zero_rates.update({t: r for t, r in pars.items() if t > max(bills.keys(), default=0)})

            sorted_tenors = sorted(combined_zero_rates.keys())
            self._zero_curves[calculation_date] = {t: combined_zero_rates[t] for t in sorted_tenors}

        curve = self._zero_curves[calculation_date]
        sorted_tenors = sorted(curve.keys())

        return ZeroCurveData(
            times=[float(t) for t in sorted_tenors],
            rates=[float(curve[t]) for t in sorted_tenors]
        )

    def bootstrap_to_zero_rates(self, par_yields: List[Tuple[float, float]]) -> Dict[float, float]:
        zero_rates = {}
        sorted_par = sorted(par_yields, key=lambda x: x[0])

        for t, c in sorted_par:
            if t <= 1.0:
                zero_rates[t] = c
            else:
                # Treasury bonds pay semi-annually. We need PV of all coupons at 0.5, 1.0, 1.5 ... t
                coupon = c / 2.0
                pv_coupons = 0.0

                # Number of semi-annual payments
                num_payments = int(round(t * 2))
                for i in range(1, num_payments):
                    payment_time = i * 0.5
                    # Interpolate the zero rate for the payment time
                    z_i = self.interpolate_zero_rate(zero_rates, payment_time)
                    pv_coupons += coupon / ((1 + z_i / 2.0) ** (2 * payment_time))

                remaining_price = 1.0 - pv_coupons
                if remaining_price <= 0:
                    zero_rates[t] = c
                    continue

                z_n = 2.0 * (((1.0 + coupon) / remaining_price) ** (1.0 / (2.0 * t)) - 1.0)
                zero_rates[t] = z_n

        return zero_rates

    def interpolate_zero_rate(self, zero_rates: Dict[float, float], t: float) -> float:
        """Linear interpolation of zero rates."""
        times = sorted(zero_rates.keys())
        if t <= times[0]:
            return zero_rates[times[0]]
        if t >= times[-1]:
            return zero_rates[times[-1]]

        # Find surrounding points
        for i in range(len(times) - 1):
            if times[i] <= t <= times[i + 1]:
                t1, t2 = times[i], times[i + 1]
                r1, r2 = zero_rates[t1], zero_rates[t2]
                return r1 + (r2 - r1) * (t - t1) / (t2 - t1)
        return zero_rates[times[-1]]

    @staticmethod
    def plot(curve: ZeroCurveData):
        import matplotlib.pyplot as plt
        plt.plot(curve.times, curve.rates)
        plt.xlabel('Time (years)')
        plt.ylabel('Zero Rate')
        plt.title('Zero Curve')
        plt.show()

if __name__ == "__main__":
    print(YieldCurve().get_zero_curve(date(2025, 12, 24)))

File: fino/src/YieldCurve.jl
module YieldCurve

#=
Description: Zero curve construction from US Treasury bill rates and par yields.
             Mirrors yield_curve.py — bootstraps par yields to zero rates,
             merges with bill rates, and caches results.
Author: seb
Date: 4/12/2026
=#

export ZeroCurveData, get_zero_curve, get_last_zero_curve,
       store_calibrated_rates, load_calibrated_rates, has_calibrated_curve

using Dates
using CSV
using DataFrames
using ..Paths

# ─────────────────────────────────────────────────────────────────────────────
# Types
# ─────────────────────────────────────────────────────────────────────────────

struct ZeroCurveData
    times ::Vector{Float32}
    rates ::Vector{Float32}
end

ZeroCurveData() = ZeroCurveData(Float32[], Float32[])

# ─────────────────────────────────────────────────────────────────────────────
# Paths
# ─────────────────────────────────────────────────────────────────────────────

function _ir_root(market::String)::String
    joinpath(PATH_DATA_INTEREST_RATE, market)
end

function _bill_rates_path(market::String, year::Int)::String
    joinpath(_ir_root(market), "daily-treasury-bill-rates", "$year.csv")
end

function _par_yield_path(market::String, year::Int)::String
    joinpath(_ir_root(market), "daily-treasury-par-yield-curve-rates", "$year.csv")
end

function _calibrated_path(ticker::String, calculation_date::Date, market::String)::String
    joinpath(_ir_root(market), "daily",
             "$(uppercase(ticker))-$(Dates.format(calculation_date, "yyyymmdd")).csv")
end

# ─────────────────────────────────────────────────────────────────────────────
# Module-level state  (mirrors Python singleton fields)
# ─────────────────────────────────────────────────────────────────────────────

# Date → Dict{Float32, Float32}  (tenor → rate)
const _bill_rates  = Dict{Date, Dict{Float32, Float32}}()
const _par_yields  = Dict{Date, Dict{Float32, Float32}}()
const _zero_curves = Dict{Date, Dict{Float32, Float32}}()
const _initialized = Set{Tuple{Date, String}}()   # (date, market)

# LRU-style caches for the two public entry points
const _zero_curve_cache      = Dict{Tuple{Date, String, String}, ZeroCurveData}()
const _last_zero_curve_cache = Dict{Tuple{Date, String, String, Int}, ZeroCurveData}()

# ─────────────────────────────────────────────────────────────────────────────
# CSV loaders
# ─────────────────────────────────────────────────────────────────────────────

function _parse_date_mdy(s::AbstractString)::Union{Date, Nothing}
    try
        return Date(s, "m/d/yyyy")
    catch
        return nothing
    end
end

function _load_bill_rates!(market::String, dt::Date)
    path = _bill_rates_path(market, year(dt))
    isfile(path) || return

    # CE columns are at 1-based indices 3,5,7,9,11,13,15 (0-based 2,4,6,8,10,12,14)
    ce_cols   = [3, 5, 7, 9, 11, 13, 15]
    tenors_f32 = Float32[4/52, 6/52, 8/52, 13/52, 17/52, 26/52, 52/52]

    open(path, "r") do f
        readline(f)   # skip header
        for line in eachline(f)
            parts = split(line, ',')
            isempty(parts[1]) && continue
            row_date = _parse_date_mdy(strip(parts[1]))
            row_date === nothing && continue

            d = get!(() -> Dict{Float32, Float32}(), _bill_rates, row_date)
            for (i, col) in enumerate(ce_cols)
                col > length(parts) && continue
                raw = strip(parts[col])
                isempty(raw) && continue
                d[tenors_f32[i]] = Float32(parse(Float64, raw) / 100.0)
            end
        end
    end
end

function _load_par_yields!(market::String, dt::Date)
    path = _par_yield_path(market, year(dt))
    isfile(path) || return

    tenors_f32 = Float32[1/12, 1.5/12, 2/12, 3/12, 4/12, 6/12,
                          1, 2, 3, 5, 7, 10, 20, 30]

    open(path, "r") do f
        readline(f)   # skip header
        for line in eachline(f)
            parts = split(line, ',')
            isempty(parts[1]) && continue
            row_date = _parse_date_mdy(strip(parts[1]))
            row_date === nothing && continue

            par_pairs = Tuple{Float32, Float32}[]
            for (i, t) in enumerate(tenors_f32)
                col = i + 1
                col > length(parts) && continue
                raw = strip(parts[col])
                isempty(raw) && continue
                push!(par_pairs, (t, Float32(parse(Float64, raw) / 100.0)))
            end

            isempty(par_pairs) && continue
            bootstrapped = _bootstrap(par_pairs)

            d = get!(() -> Dict{Float32, Float32}(), _par_yields, row_date)
            merge!(d, bootstrapped)
        end
    end
end

function _initialize!(market::String, dt::Date)
    key = (dt, market)
    key in _initialized && return
    push!(_initialized, key)
    _load_bill_rates!(market, dt)
    _load_par_yields!(market, dt)
end

# ─────────────────────────────────────────────────────────────────────────────
# Bootstrap par → zero  (semi-annual Treasury convention)
# ─────────────────────────────────────────────────────────────────────────────

function _interpolate_zero(zero_rates::Dict{Float32, Float32}, t::Float32)::Float32
    times = sort(collect(keys(zero_rates)))
    isempty(times) && return 0f0
    t <= times[1]   && return zero_rates[times[1]]
    t >= times[end] && return zero_rates[times[end]]
    for i in 1:length(times)-1
        if times[i] <= t <= times[i+1]
            t1, t2 = times[i], times[i+1]
            r1, r2 = zero_rates[t1], zero_rates[t2]
            return r1 + (r2 - r1) * (t - t1) / (t2 - t1)
        end
    end
    return zero_rates[times[end]]
end

function _bootstrap(par_pairs::Vector{Tuple{Float32, Float32}})::Dict{Float32, Float32}
    zero_rates = Dict{Float32, Float32}()
    sort!(par_pairs, by=first)

    for (t, c) in par_pairs
        if t <= 1f0
            zero_rates[t] = c
        else
            coupon       = c / 2f0
            pv_coupons   = 0f0
            num_payments = round(Int, t * 2)

            for i in 1:num_payments-1
                payment_time = Float32(i) * 0.5f0
                z_i = _interpolate_zero(zero_rates, payment_time)
                pv_coupons += coupon / (1f0 + z_i / 2f0) ^ (2f0 * payment_time)
            end

            remaining = 1f0 - pv_coupons
            if remaining <= 0f0
                zero_rates[t] = c
                continue
            end

            z_n = 2f0 * (((1f0 + coupon) / remaining) ^ (1f0 / (2f0 * t)) - 1f0)
            zero_rates[t] = z_n
        end
    end

    return zero_rates
end

# ─────────────────────────────────────────────────────────────────────────────
# Calibrated curve persistence
# ─────────────────────────────────────────────────────────────────────────────

function has_calibrated_curve(calculation_date::Date, ticker::String, market::String="usa")::Bool
    isfile(_calibrated_path(ticker, calculation_date, market))
end

function store_calibrated_rates(ticker::String, calculation_date::Date,
                                curve::ZeroCurveData, market::String="usa")
    path = _calibrated_path(ticker, calculation_date, market)
    mkpath(dirname(path))
    df = DataFrame(Tenor=Float64.(curve.times), Rate=Float64.(curve.rates))
    CSV.write(path, df)
end

function load_calibrated_rates(ticker::String, calculation_date::Date,
                               market::String="usa")::ZeroCurveData
    path = _calibrated_path(ticker, calculation_date, market)
    isfile(path) || error("load_calibrated_rates(): $path does not exist.")
    df = CSV.read(path, DataFrame)
    ZeroCurveData(Float32.(df.Tenor), Float32.(df.Rate))
end

# ─────────────────────────────────────────────────────────────────────────────
# Core curve construction
# ─────────────────────────────────────────────────────────────────────────────

function _get_latest_available_curve(from_date::Date, market::String,
                                     max_days::Int=180)::ZeroCurveData
    for offset in 1:max_days
        candidate = from_date - Day(offset)
        _initialize!(market, candidate)
        bills = get(_bill_rates, candidate, Dict{Float32, Float32}())
        pars  = get(_par_yields,  candidate, Dict{Float32, Float32}())
        if !isempty(bills) || !isempty(pars)
            @info "get_zero_curve(): rates unavailable for $from_date, using $candidate instead"
            combined = Dict{Float32, Float32}(bills)
            bill_max = isempty(bills) ? 0f0 : maximum(keys(bills))
            for (t, r) in pars
                t > bill_max && (combined[t] = r)
            end
            sorted_tenors = sort(collect(keys(combined)))
            return ZeroCurveData(sorted_tenors, [combined[t] for t in sorted_tenors])
        end
    end
    return ZeroCurveData()
end

"""
    get_zero_curve(calculation_date, ticker="", market="usa") -> ZeroCurveData

Returns the zero curve for `calculation_date`.  If a calibrated curve for
`ticker` exists on disk it is loaded directly; otherwise the curve is built
from bill rates + bootstrapped par yields.
"""
function get_zero_curve(calculation_date::Date,
                        ticker::String="",
                        market::String="usa")::ZeroCurveData
    cache_key = (calculation_date, ticker, market)
    haskey(_zero_curve_cache, cache_key) && return _zero_curve_cache[cache_key]

    # Prefer calibrated curve when ticker is provided
    if !isempty(ticker) && has_calibrated_curve(calculation_date, ticker, market)
        result = load_calibrated_rates(ticker, calculation_date, market)
        _zero_curve_cache[cache_key] = result
        return result
    end

    _initialize!(market, calculation_date)

    if !haskey(_zero_curves, calculation_date)
        bills = get(_bill_rates, calculation_date, Dict{Float32, Float32}())
        pars  = get(_par_yields,  calculation_date, Dict{Float32, Float32}())

        if isempty(bills) && isempty(pars)
            # Rates not yet published for this date — fall back to the most
            # recent date for which we do have data (up to 14 calendar days back).
            fallback_result = _get_latest_available_curve(calculation_date, market)
            _zero_curve_cache[cache_key] = fallback_result
            return fallback_result
        end

        combined = Dict{Float32, Float32}(bills)

        bill_max = isempty(bills) ? 0f0 : maximum(keys(bills))
        for (t, r) in pars
            t > bill_max && (combined[t] = r)
        end

        sorted_tenors = sort(collect(keys(combined)))
        _zero_curves[calculation_date] = Dict(t => combined[t] for t in sorted_tenors)
    end

    curve = _zero_curves[calculation_date]
    sorted_tenors = sort(collect(keys(curve)))
    result = ZeroCurveData(sorted_tenors, [curve[t] for t in sorted_tenors])

    _zero_curve_cache[cache_key] = result
    return result
end

"""
    get_last_zero_curve(calculation_date, ticker="", delta_days=-14, market="usa") -> ZeroCurveData

Searches backwards from `calculation_date + delta_days` to `calculation_date`
for the most recent calibrated curve for `ticker`. Falls back to the generic
zero curve if none is found.
"""
function get_last_zero_curve(calculation_date::Date,
                             ticker::String="",
                             delta_days::Int=-14,
                             market::String="usa")::ZeroCurveData
    cache_key = (calculation_date, ticker, market, delta_days)
    haskey(_last_zero_curve_cache, cache_key) && return _last_zero_curve_cache[cache_key]

    # Walk backwards over the window [calculation_date + delta_days .. calculation_date]
    start_date = calculation_date + Day(delta_days)   # delta_days is negative → earlier date
    for offset in 0:-sign(delta_days):delta_days
        candidate = calculation_date + Day(offset)
        candidate < start_date && break
        if !isempty(ticker) && has_calibrated_curve(candidate, ticker, market)
            @info "get_last_zero_curve(): using calibrated curve" calculation_date candidate ticker
            result = get_zero_curve(candidate, ticker, market)
            _last_zero_curve_cache[cache_key] = result
            return result
        end
    end

    # Fallback: generic treasury zero curve for calculation_date
    result = get_zero_curve(calculation_date, "", market)
    _last_zero_curve_cache[cache_key] = result
    return result
end

function get_last_zero_curve(calculation_date::Date, ticker, delta_days::Int=-14, market::String="usa")::ZeroCurveData
    get_last_zero_curve(calculation_date, uppercase(ticker.symbol), delta_days, market)
end

end # module YieldCurve